tests/reconfiguration.py

```diff
@@ -610,6 +611,106 @@ def run_join_old_snapshot(args):
+                for nid, info_ in nodes.items():
+                    info = json.loads(info_)
+                    if nid.decode() == learner_id and "status" in info:
+                        seq_no = tr.get_public_domain().get_seqno()
+                        if info["status"] == "Pending":
+                            pending_at = seq_no
+                        elif info["status"] == "Learner":
+                            learner_at = seq_no
+                        elif info["status"] == "Trusted":
+                            trusted_at = seq_no
+
+    assert pending_at < learner_at < trusted_at
+
+
+def run_migration_tests(args):
+    if args.reconfiguration_type != "OneTransaction":
+        return
+
+    txs = app.LoggingTxs("user0")
+    ledger_paths = None
+    learner_id = None
+
+    with infra.network.network(
+        args.nodes,
+        args.binary_dir,
+        args.debug_nodes,
+        args.perf_nodes,
+        pdb=args.pdb,
+        txs=txs,
+    ) as network:
+        network.start_and_join(args)
+        primary, _ = network.find_primary()
+
+        nodes_before = get_current_nodes_table(network)
+        for n in network.nodes:
+            assert n.node_id in nodes_before
+            assert nodes_before[n.node_id]["status"] == "Trusted"
+
+        # Check that the service config agrees that this is a 1tx network
+        with primary.client() as c:
+            s = c.get("/node/service/configuration").body.json()
+            assert s["reconfiguration_type"] == "OneTransaction"
+
+        network.consortium.submit_2tx_migration_proposal(primary)
+        network.wait_for_all_nodes_to_commit(primary)
+
+        # Check that the service config has been updated
+        with primary.client() as c:
+            rj = c.get("/node/service/configuration").body.json()
+            assert rj["reconfiguration_type"] == "TwoTransaction"
+
+        # Check that all nodes have updated their consensus parameters
+        for node in network.nodes:
+            with node.client() as c:
+                rj = c.get("/node/consensus").body.json()
+                assert "reconfiguration_type" in rj["details"]
+                assert rj["details"]["reconfiguration_type"] == "TwoTransaction"
+                assert len(rj["details"]["learners"]) == 0
+
+        test_add_node(network, args)
+        new_node = network.nodes[-1]
+
+        # Check that the new node has the right consensus parameter
+        with new_node.client() as c:
+            rj = c.get("/node/consensus").body.json()
+            assert "reconfiguration_type" in rj["details"]
+            assert "learners" in rj["details"]
+            assert rj["details"]["reconfiguration_type"] == "TwoTransaction"
+            assert len(rj["details"]["learners"]) == 0
+
+        ledger_paths = primary.remote.ledger_paths()
+        learner_id = new_node.node_id
+
+        # Force ledger flush of all transactions so far.
+        network.get_latest_ledger_public_state()
```

Comment:
@jumaffre this is for the new checks on the ledger files. Is there a better/nicer way to force the ledger files to be flushed to disk? (It's necessary, I don't see the 2nd reconfig tx without it.)

Reply:
Ah yes, this is because you passed `committed_only=True` when creating the `Ledger` instance on line `628` (which means only `.committed` ledger files are considered). Because you call `check_2tx_ledger()` when the network is stopped, all ledger files should be neatly closed and it should be OK to use `committed_only=False` and avoid having to call `network.get_latest_ledger_public_state()` here.

Reply:
Thanks, that's great, looks much neater now!